maven-resolver-util/src/main/java/org/eclipse/aether/util/FileUtils.java

```
@@ -98,23 +108,59 @@ public static CollocatedTempFile newTempFile(Path file) throws IOException {
         Path tempFile = parent.resolve(file.getFileName() + "."
                 + Long.toUnsignedString(ThreadLocalRandom.current().nextLong()) + ".tmp");
         return new CollocatedTempFile() {
+            private final AtomicBoolean wantsMove = new AtomicBoolean(false);
+
             @Override
             public Path getPath() {
                 return tempFile;
             }
 
             @Override
-            public void move() throws IOException {
-                Files.move(tempFile, file, StandardCopyOption.ATOMIC_MOVE);
+            public void move() {
+                wantsMove.set(true);
             }
 
             @Override
             public void close() throws IOException {
+                if (wantsMove.get() && Files.isReadable(tempFile)) {
+                    fsyncFile(tempFile);
+                    Files.move(tempFile, file, StandardCopyOption.ATOMIC_MOVE);
+                    if (!IS_WINDOWS) {
+                        fsyncParent(tempFile);
+                    }
+                }
                 Files.deleteIfExists(tempFile);
             }
         };
     }
 
+    /**
+     * Performs fsync: makes sure no OS "dirty buffers" exist for given file.
+     *
+     * @param target Path that must not be {@code null}, must exist as plain file.
+     */
+    private static void fsyncFile(Path target) throws IOException {
+        try (FileChannel file = FileChannel.open(target, StandardOpenOption.WRITE)) {
+            file.force(true);
+        }
```

should we also ignore exception here as is in `fsyncParent`?

---

Lucene code in case of file does not ignore exception, so I think we should not either.